### PR TITLE
Crimson: Support basic deployments

### PR DESCRIPTION
**Pull request by Matan-B**
This new option is there to allow *basic* cluster deployments.
crimson_seastar_num_threads can be used globally for all the OSDs in the cluster:
```
osd:
  crimson_seastar_num_threads: <n>
  crimson_alien_op_num_threads: <m>
```

As a result, all the available CPUs will be allocated to *both* seastar reactor threads
and to Alienstore threads - without any exclusion.

Notes:

* The core allocation will most likely overlap between OSDS and/or Seastar and Alienstore.
An optiomal deployment requires `crimson_alien_thread_cpu_cores`
and `crimson_seastar_cpu_cores` to be set for each OSD based on the host its located at.

* Seastar reactor number (smp::count) will be deduced directly from crimson_seastar_num_threads

---

Documentation followup: https://github.com/ceph/ceph/pull/57352

---
### Option 1 (Optimal)
* CPUs are pinned to the provided sets (seastar and alien cpu sets)
* thread-affinity = 1 (seastar's default).
* smp (smp::count / reactor count): `seastar_cpu_set.size()`
* Seastar reactor threads: `seastar_cpu_set.size()`
* Alienstore threads num: `crimson_alien_op_num_threads`
---
### Option 2 (Basic):
* No CPU pinning.
* thread-affinity = 0
* smp (smp::count / reactor count) = `crimson_seastar_num_threads`
* Seastar reactor threads num: `crimson_seastar_num_threads`
* Alienstore threads num: `crimson_alien_op_num_threads`
---
Fixes: https://tracker.ceph.com/issues/65752

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/common/options/crimson.yaml.in [hunk 1]
```diff
@@ -31,6 +31,13 @@ options:
   desc: CPU cores on which alienstore threads will run in cpuset(7) format
   flags:
   - startup
+- name: crimson_seastar_reactor_num
+  type: uint
+  level: advanced
+  default: 0
+  desc: number of CPU cores on which seastar reactor threads will run
+  flags:
+  - startup
```
Comment (cyx1231st) [inline src/common/options/crimson.yaml.in:40]: Not sure why `crimson_seastar_reactor_num` is needed.

Looks it is identical to configure `crimson_seastar_cpu_cores` as "0-`crimson_seastar_reactor_num-1`" ?

Which seems not useful when there are multiple OSDs because `--cpuset` always start with 0, causing overlap.

Comment (Matan-B) [reply src/common/options/crimson.yaml.in:40]: I agree that it's practically the same. It's there only to make Crimson deployment friendlier by setting a global (non-optimal) OSD value of `crimson_seastar_reactor_num = n`.
Once `crimson_seastar_cpu_cores` and `crimson_alien_thread_cpu_cores` are set, this value will no longer have any impact.

Another reason is that if `crimson_seastar_cpu_cores` is not set, we fallback to Seastar's default of using all available cores. That means that on nodes with 80 cores we will create 80 seastar threads.

By looking again, `crimson_seastar_thread_num` is probably a better terminology.

Comment (cyx1231st) [reply src/common/options/crimson.yaml.in:40]: I think I generally understand the purpose.

> Which seems not useful when there are multiple OSDs because --cpuset always start with 0, causing overlap.

But I'm still wondering how to prevent "--cpuset" overlapping if there are multiple OSDs in the same host, only with `crimson_seastar_thread_num` or `crimson_seastar_reactor_num` configured.

I suppose the OSD still needs to know that it is the nth daemon in the same host to produce the accurate `--cpuset` for seastar. But if it requires yet another Ceph configuration (which still has to be configured separately), probably the current `crimson_seastar_cpu_cores` is simpler?

Or, does each OSD live in an separate container, so that overlapping is impossible?

Comment (Matan-B) [reply src/common/options/crimson.yaml.in:40]: > But I'm still wondering how to prevent "--cpuset" overlapping if there are multiple OSDs in the same host, only with `crimson_seastar_thread_num` or `crimson_seastar_reactor_num` configured.

It's expected to overlap. As mentioned, it's not optimal (at all), but it's just there to let the cluster come up in a simple way. Actual usage will have to set `crimson_seastar_cpu_cores` and `crimson_alien_thread_cpu_cores`.
I'll emphasis that in the docs.

Comment (Matan-B) [reply src/common/options/crimson.yaml.in:40]: @cyx1231st, Please see updated commit

Comment (athanatos) [reply src/common/options/crimson.yaml.in:40]: The difference is that if all OSDs on a host specify 0-N for crimson_sesatar_cpu_cores they will all be limited to exactly the same set of cpus -- that's basically always undesirable.  With crimson_seastar_reactor_num, we don't set a cpu_mask at all so the scheduler is free to place them whereever.  In that case, they *may* overlap, but at least all of the cpus on the host will be used.  Operators interested in performance will need to specify crimson_seastar_cpu_cores for every OSD individually, but operators who simply want to try to it out can specify crimson_seastar_reactor_num a single time globally and get reasonable if suboptimal performance.  Such operators can then choose to switch per-osd crimson_seastar_cpu_cores if they wish.

Perhaps it would make sense to send a warning to the central log if a crimson-osd starts without crimson_seastar_cpu_cores (and the corresponding bluestore ones for bluestore) set to alert operators that they may want to make changes for better performance?

Comment (cyx1231st) [reply src/common/options/crimson.yaml.in:40]: @athanatos yeah, makes sense to me.

Probably `--smp` and `--thread-affinity=0` can be useful for `crimson_seastar_num_threads`.

The same warning as well as how the 3 configs impact alienstore thread pinning should also be good in the configuration description. User may not look at the comments in the source code.

Comment (Matan-B) [reply src/common/options/crimson.yaml.in:40]: >Perhaps it would make sense to send a warning to the central log if a crimson-osd starts without crimson_seastar_cpu_cores (and the corresponding bluestore ones for bluestore) set to alert operators that they may want to make changes for better performance?

@athanatos, Does that qualify as a health check alert?

---

>Probably --smp and --thread-affinity=0 can be useful for crimson_seastar_num_threads.

* @cyx1231st, What do you suggest to change with `--smp`? Currently we only allow 1 (seastar`s default of one thread per CPU).

* `--thread-affinity=0` may be a good idea for  `crimson_seastar_num_threads`.

>The same warning as well as how the 3 configs impact alienstore thread pinning should also be good in the configuration description. User may not look at the comments in the source code.

Ack, will update the desc. Docs will be updated in a followup PR.

Comment (athanatos) [reply src/common/options/crimson.yaml.in:40]: @Matan-B No, we don't want to force users to deal with it, merely notify them.  Perhaps just printing a local log message at info would be fine?

Comment (cyx1231st) [reply src/common/options/crimson.yaml.in:40]: `-c [ --smp ] arg                      number of threads (default: one per CPU)`

Looks it configures the seastar thread number, and `thread-affinity=0` to unpin the CPUs.

Comment (Matan-B) [reply src/common/options/crimson.yaml.in:40]: See reactor.cc:
```
    if (smp_opts.smp) {
        smp::count = smp_opts.smp.get_value();
    } else {
        smp::count = cpu_set.size();
    }
    std::vector<reactor*> reactors(smp::count);
```
All that `--smp` does is replace cpu_set.size() usage explicitly. Setting `--smp` = n or `--cpuset` = 0-n-1 will result in the **same** number of smp::count. The only difference would be which CPUs are used.

Comment (athanatos) [reply src/common/options/crimson.yaml.in:40]: I agree that it makes sense to allow the user to specify a number of alienstore cpu threads different from the number of seastar threads.  To make things simpler, I'd be fine with requiring the user to specify both.

Comment (Matan-B) [reply src/common/options/crimson.yaml.in:47]: ### Option 1 (Optimal)
* CPUs are pinned to the provided sets (seastar and alien cpu sets)
* thread-affinity = 1 (seastar's default).
* smp (smp::count / reactor count): `seastar_cpu_set.size()`
* Seastar reactor threads: `seastar_cpu_set.size()`
* Alienstore threads: `alien_cpu_set.size()` * num_threads_per_core (default = 4).
--- 
### Option 2 (Basic):
* No CPU pinning.
* thread-affinity = 0
* smp (smp::count / reactor count) = `crimson_seastar_num_threads`
* Seastar reactor threads num: `crimson_seastar_num_threads`
* Alienstore threads num: `crimson_alien_op_num_threads`

File: src/common/options/crimson.yaml.in [hunk 2]
```diff
@@ -16,7 +16,14 @@ options:
...
   desc: The number of threads for serving alienized ObjectStore
-  default: 6
+  default: 0
+  flags:
+  - startup
+- name: crimson_alien_op_num_threads_per_core
+  type: uint
+  level: advanced
+  desc: The number of threads (per core) for serving alienized ObjectStore
+  default: 4
```
Comment (cyx1231st) [inline src/common/options/crimson.yaml.in:26]: The existing `crimson_alien_op_num_threads` can already set the threads of alienstore workers.

`crimson_alien_op_num_threads_per_core` works as a suger to set it by multiplying `alien_cpu_set.size()`, but it might not be really useful (or even misleading) from my observations:

We tried to find the optimal configuration using different kinds of CPUs (different core numbers, different per-core performance), by tuning the core allocation ratio of seastar reactor vs alienstore, and the number of alienstore threads. But seems the configuration of `num-alien-op-threads` is independent to the `alien-cpu-set`. For example, in one platform, the ratio is 4, and in another platform, the ratio turns out to be 1. IIUC exposing an additional `crimson_alien_op_num_threads_per_core` might be misleading to suggest that there is a preferred default.

And I think `crimson_alien_op_num_threads_per_core (4)` is also misleading when there are multiple OSDs sharing the same `alien_cpu_set`, so that the actual running alienstore threads per core is 8 if there are 2 OSDs, and 16 when there are 4 OSDs.

In short, seems to me the original `crimson_alien_op_num_threads` is also sufficient for Option 1 (Optimal).

File: src/crimson/osd/main_config_bootstrap_helpers.cc [hunk 3]
```diff
@@ -150,16 +150,32 @@ _get_early_config(int argc, const char *argv[])
...
+	    auto reactor_num = crimson::common::get_conf<uint64_t>("crimson_seastar_num_threads");
+	    if (!reactor_num) {
+	      logger().error("get_early_config: crimson_seastar_cpu_cores"
+                             " or crimson_seastar_num_threads"
+                             " must be set");
+	      ceph_abort();
+	    }
+	    std::string smp = fmt::format("{}", reactor_num);
+	    ret.early_args.emplace_back("--smp");
+	    ret.early_args.emplace_back(smp);
```
Comment (cyx1231st) [inline src/crimson/osd/main_config_bootstrap_helpers.cc:173]: Ack.

File: src/common/options/crimson.yaml.in (hunk 2 above)
Comment (Matan-B) [reply src/common/options/crimson.yaml.in:26]: I agree that "per-core" might be misleading. Thank you for the review, I'll go back to only having num_threads.

Comment (Matan-B) [reply src/common/options/crimson.yaml.in]: ### Option 1 (Optimal)
...
* ~Alienstore threads: `alien_cpu_set.size()` * `crimson_alien_op_num_threads_per_core` (default = 4).~
* Alienstore threads num: `crimson_alien_op_num_threads`

...

**Comment by Matan-B** [comment]
@cyx1231st, @athanatos - should be ready with all comments addressed.

File: src/common/options/crimson.yaml.in [hunk 4]
```diff
@@ -31,6 +31,15 @@ options:
   desc: CPU cores on which alienstore threads will run in cpuset(7) format
   flags:
   - startup
+- name: crimson_seastar_num_threads
+  type: uint
+  level: advanced
+  default: 0
+  desc: The number of threads for serving seastar reactors
```
Comment (cyx1231st) [inline src/common/options/crimson.yaml.in:38]: Can be more informative:

The number of threads for serving seastar reactors without CPU pinning, won't be effective if crimson_seastar_cpu_cores is set.

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-june-9-E/

**Comment by Matan-B** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by Matan-B** [comment]
Merging based on (21/21 passed):
https://pulpito.ceph.com/matan-2024-06-09_12:08:27-crimson-rados-wip-matanb-crimson-testing-june-9-E-distro-crimson-smithi/
Rereun dead job:
https://pulpito.ceph.com/matan-2024-06-09_13:09:45-crimson-rados-wip-matanb-crimson-testing-june-9-E-distro-crimson-smithi/

**Comment by Matan-B** [comment]
jenkins test api
